third_party/terraform/tests/resource_container_node_pool_test.go.erb

```diff
@@ -242,6 +242,37 @@ func TestAccContainerNodePool_withSandboxConfig(t *testing.T) {
 		},
 	})
 }
+
+func TestAccContainerNodePool_withUpgradeSettings(t *testing.T) {
+	t.Parallel()
+
+	clusterName := acctest.RandString(10)
+	nodePoolName := acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckContainerClusterDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccContainerNodePool_withUpgradeSettings(clusterName, nodePoolName, 2, 3),
+			},
+			{
+				ResourceName:      "google_container_node_pool.with_upgrade_settings",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccContainerNodePool_withUpgradeSettings(clusterName, nodePoolName, 1, 1),
```

Comment:
Can you add a step where the values are changed to zero? My guess right now is that it won't work, and that you'll be able to fix it by adding `ForceSendFields:  []string{"MaxSurge", "MaxUnavailable"}` to your requests, but it'll be good to verify.

Reply:
While addressing this comment, I found out that the API requires that max_surge and max_unavailable must not be negative and at least one of them must be greater than zero. 

This requires special validation logic as the `ValidateFunc` (https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#validatefunc) does not support it.

Reply:
I am handling this case with a function https://github.com/GoogleCloudPlatform/magic-modules/pull/2669/files#diff-0b06c952a9086a32669348914ce0f5b6R486-R491 that will be checked during resource creation and update.